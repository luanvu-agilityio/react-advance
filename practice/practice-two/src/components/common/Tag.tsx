import type { MouseEvent, ReactNode, CSSProperties } from 'react'
import styled from 'styled-components'
import Link from './Link'

// Base props for all tag variants
type BaseTagProps = {
  label: string
  variant?: 'default' | 'selected' | 'disabled'
  className?: string
  style?: CSSProperties
  children?: ReactNode
}

// Props specific to button variant
type ButtonTagProps = BaseTagProps & {
  as: 'button'
  onClick: (e: MouseEvent<HTMLButtonElement>) => void
  disabled?: boolean
}

// Props specific to link variant
type LinkTagProps = BaseTagProps & {
  as: 'link'
  href: string
  target?: '_blank' | '_self' | '_parent' | '_top'
  onClick: (e: MouseEvent<HTMLAnchorElement>) => void
}

export type TagProps = ButtonTagProps | LinkTagProps

const StyledButtonTag = styled.button<{
  $variant?: 'default' | 'selected' | 'disabled'
}>`
  display: inline-flex;
  align-items: center;
  padding: 4px 10px;
  border-radius: 12px;
  font-size: 12px;
  font-weight: var(--font-weight-semibold);
  transition: all 0.2s ease;
  border: none;
  cursor: ${({ $variant }) =>
    $variant === 'disabled' ? 'not-allowed' : 'pointer'};

  /* Variant styles */
  ${({ $variant }) => {
    switch ($variant) {
      case 'selected':
        return `
            background-color:var(--cyan-shade-4);
            color: var(--cyan-shade-1);
            box-shadow: 0 0 0 2px #a8c7fa;
            &:hover {
              background-color:var(--cyan-shade-3);
              text-decoration: none;
          }
        `
      case 'disabled':
        return `
          background-color: var(--black-shade-6);
          color:var(--black-shade-3);
          opacity: 0.7;
        `
      default:
        return `
          background-color: var(--black-shade-6);
          color: var(--black-color-default);
          &:hover {
            background-color: var(--black-shade-5);
          }
        `
    }
  }}
`

const StyledLinkWrapper = styled.span<{
  $variant?: 'default' | 'selected' | 'disabled'
}>`
  display: inline-block;

  /* Reset Link styles when used in a tag */
  a {
    display: inline-flex;
    align-items: center;
    padding: 4px 10px;
    border-radius: 12px;
    font-size: 12px;
    font-weight: 600;
    transition: all 0.2s ease;
    text-decoration: none;

    /* Variant styles */
    ${({ $variant }) => {
      switch ($variant) {
        case 'selected':
          return `
            background-color: var(--green-shade-3);
            color:var(--green-shade-1);
            box-shadow: 0 0 0 2px #a8c7fa;
            &:hover {
              background-color:var(--green-shade-2);
              text-decoration: none;
            }
          `
        case 'disabled':
          return `
         background-color: var(--black-shade-6);
          color:var(--black-shade-3);
          opacity: 0.7;
          `
        default:
          return `
            background-color: var(--black-shade-5);
            &:hover {
              background-color: var(--black-shade-6);
              text-decoration: none;
            }
          `
      }
    }}
  }
`

export const Tag = (props: TagProps) => {
  const { label, variant = 'default', className = '', style } = props

  // For button variant
  if (props.as === 'button') {
    return (
      <StyledButtonTag
        type="button"
        className={className}
        $variant={variant}
        onClick={props.onClick}
        disabled={props.disabled ?? variant === 'disabled'}
        style={style}
      >
        {props.children ?? label}
      </StyledButtonTag>
    )
  }

  return (
    <StyledLinkWrapper $variant={variant} className={className}>
      <Link
        href={props.href}
        onClick={props.onClick}
        target={props.target ?? '_self'}
        disabled={variant === 'disabled'}
        style={style}
      >
        {props.children ?? label}
      </Link>
    </StyledLinkWrapper>
  )
}

export default Tag
