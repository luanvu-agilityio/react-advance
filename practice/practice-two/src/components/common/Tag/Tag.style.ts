import styled from 'styled-components'

export const StyledButtonTag = styled.button<{
  $variant?: 'default' | 'selected' | 'disabled'
}>`
  display: inline-flex;
  align-items: center;
  padding: 4px 10px;
  border-radius: 12px;
  font-size: 12px;
  font-weight: var(--font-weight-semibold);
  transition: all 0.2s ease;
  border: none;
  cursor: ${({ $variant }) =>
    $variant === 'disabled' ? 'not-allowed' : 'pointer'};

  /* Variant styles */
  ${({ $variant }) => {
    switch ($variant) {
      case 'selected':
        return `
            background-color:var(--cyan-shade-4);
            color: var(--cyan-shade-1);
            box-shadow: 0 0 0 2px #a8c7fa;
            &:hover {
              background-color:var(--cyan-shade-3);
              text-decoration: none;
          }
        `
      case 'disabled':
        return `
          background-color: var(--black-shade-6);
          color:var(--black-shade-3);
          opacity: 0.7;
        `
      default:
        return `
          background-color: var(--black-shade-5);
          color: var(--black-color-default);
          &:hover {
            background-color: var(--black-shade-5);
          }
        `
    }
  }}

  &[data-tag-type="category"] {
    border-color: var(--green-color-default);
    border-width: 2px;

    &:hover {
      background-color: var(--green-50);
      border-color: var(--green-color-dark);
    }
  }

  &[data-tag-type='subcategory'] {
    border-color: var(--blue-color-default);
    border-width: 2px;

    &:hover {
      background-color: var(--blue-50);
      border-color: var(--blue-color-dark);
    }
  }

  &[data-tag-type='product'] {
    border-color: var(--gray-300);
    border-width: 1px;

    &:hover {
      background-color: var(--gray-50);
      border-color: var(--gray-400);
    }
  }
`

export const StyledLinkWrapper = styled.span<{
  $variant?: 'default' | 'selected' | 'disabled'
}>`
  display: inline-block;

  /* Reset Link styles when used in a tag */
  a {
    display: inline-flex;
    align-items: center;
    padding: 4px 10px;
    border-radius: 12px;
    font-size: 12px;
    font-weight: 600;
    transition: all 0.2s ease;
    text-decoration: none;

    /* Variant styles */
    ${({ $variant }) => {
      switch ($variant) {
        case 'selected':
          return `
            background-color: var(--green-shade-3);
            color:var(--green-shade-1);
            box-shadow: 0 0 0 2px #a8c7fa;
            &:hover {
              background-color:var(--green-shade-2);
              text-decoration: none;
            }
          `
        case 'disabled':
          return `
         background-color: var(--black-shade-6);
          color:var(--black-shade-3);
          opacity: 0.7;
          `
        default:
          return `
            background-color: var(--black-shade-5);
            &:hover {
              background-color: var(--black-shade-6);
              text-decoration: none;
            }
          `
      }
    }}
  }
`
