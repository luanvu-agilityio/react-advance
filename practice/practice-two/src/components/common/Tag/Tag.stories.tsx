import type { Meta, StoryObj } from '@storybook/react'
import styled from 'styled-components'
import Tag from './index'

const TagsContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  gap: 0.5rem;
`

const CategoryContainer = styled.div`
  .fruit a {
    background-color: var(--green-shade-3);
    color: var(--green-shade-1);
    &:hover {
      background-color: var(--green-shade-2);
    }
  }

  .vegetable a {
    background-color: var(--coral-shade-3);
    color: var(--coral-shade-1);
    &:hover {
      background-color: var(--coral-shade-2);
    }
  }

  .protein a {
    background-color: var(--brown-shade-3);
    color: var(--brown-shade-1);
    &:hover {
      background-color: var(--brown-shade-2);
    }
  }
`

const meta = {
  title: 'Components/Common/Tag',
  component: Tag,
  parameters: {
    layout: 'padded',
  },
  tags: ['autodocs'],
  argTypes: {
    as: {
      control: 'radio',
      options: ['button', 'link'],
    },
    variant: {
      control: 'radio',
      options: ['default', 'selected', 'disabled'],
    },
    label: { control: 'text' },
    onClick: { action: 'clicked' },
    href: { control: 'text' },

    className: { control: 'text' },
    disabled: { control: 'boolean' },
  },
} satisfies Meta<typeof Tag>

export default meta
type Story = StoryObj<typeof Tag>

// Basic Button Tag
export const ButtonTag: Story = {
  args: {
    as: 'button',
    label: 'Button Tag',
    onClick: () => console.log('Clicked'),
  },
}

// Selected Button Tag
export const SelectedButtonTag: Story = {
  args: {
    as: 'button',
    label: 'Selected Tag',
    variant: 'selected',
    onClick: () => console.log('Clicked'),
  },
}

// Disabled Button Tag
export const DisabledButtonTag: Story = {
  args: {
    as: 'button',
    label: 'Disabled Tag',
    variant: 'disabled',
    onClick: () => console.log('Clicked'),
  },
}

// Basic Link Tag
export const LinkTag: Story = {
  args: {
    as: 'link',
    label: 'Link Tag',
    href: '#',
    onClick: (e) => e.preventDefault(),
  },
}

// Multiple Tags Group
export const TagGroup: Story = {
  decorators: [
    () => (
      <TagsContainer>
        <Tag as="button" label="Beans" onClick={() => {}} />
        <Tag
          as="button"
          label="Carrots"
          variant="selected"
          onClick={() => {}}
        />
        <Tag as="button" label="Apples" onClick={() => {}} />
        <Tag
          as="button"
          label="Disabled"
          variant="disabled"
          onClick={() => {}}
        />
      </TagsContainer>
    ),
  ],
}

// Category Link Tags
export const CategoryLinkTags: Story = {
  decorators: [
    () => (
      <CategoryContainer>
        <TagsContainer>
          <div className="fruit">
            <Tag
              as="link"
              label="Fruits"
              href="#"
              onClick={(e) => e.preventDefault()}
            />
          </div>
          <div className="vegetable">
            <Tag
              as="link"
              label="Vegetables"
              href="#"
              onClick={(e) => e.preventDefault()}
            />
          </div>
          <div className="protein">
            <Tag
              as="link"
              label="Meat & Fish"
              href="#"
              onClick={(e) => e.preventDefault()}
            />
          </div>
        </TagsContainer>
      </CategoryContainer>
    ),
  ],
}

// Interactive Tag Group
export const InteractiveTagGroup: Story = {
  parameters: {
    docs: {
      story: { inline: true },
    },
  },
  render: function Render() {
    const productTags = [
      'Beans',
      'Carrots',
      'Apples',
      'Garlic',
      'Mushrooms',
      'Tomatoes',
    ]

    return (
      <TagsContainer>
        {productTags.map((tag) => (
          <Tag
            key={tag}
            as="button"
            label={tag}
            onClick={() => console.log(`Clicked ${tag}`)}
          />
        ))}
      </TagsContainer>
    )
  },
}
