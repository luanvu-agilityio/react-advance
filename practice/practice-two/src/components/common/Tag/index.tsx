import type { ButtonHTMLAttributes, ReactNode } from 'react'
import Link from '@components/common/Link'
import { StyledButton, StyledContainer } from './Tag.style'

type TagVariant = 'default' | 'selected' | 'disabled'
type TagAs = 'button' | 'link'

interface BaseTagProps {
  variant?: TagVariant
  label?: string
  className?: string
  children?: ReactNode
  as?: TagAs
}

interface ButtonTagProps
  extends BaseTagProps,
    ButtonHTMLAttributes<HTMLButtonElement> {
  as: 'button'
  onClick: () => void
  href?: never
}

interface LinkTagProps extends BaseTagProps {
  as: 'link'
  href: string
  onClick?: (e: React.MouseEvent<HTMLAnchorElement>) => void
}

type TagProps = ButtonTagProps | LinkTagProps

const Tag = ({
  variant = 'default',
  label = '',
  className,
  children,
  as = 'button',
  ...props
}: TagProps) => {
  const isDisabled = variant === 'disabled'

  // Render either button or link based on 'as' prop
  if (as === 'button') {
    const { onClick } = props as ButtonTagProps
    return (
      <StyledButton
        $variant={variant}
        onClick={isDisabled ? undefined : onClick}
        disabled={isDisabled}
        className={className}
      >
        {children || label}
      </StyledButton>
    )
  } else {
    // For link variant
    const { href, onClick } = props as LinkTagProps
    return (
      <StyledContainer $variant={variant} className={className}>
        <Link href={href} onClick={onClick} disabled={isDisabled}>
          {children ?? label}
        </Link>
      </StyledContainer>
    )
  }
}

export default Tag
