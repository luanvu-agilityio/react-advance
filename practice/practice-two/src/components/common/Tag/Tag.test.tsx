import { render, screen, fireEvent } from '@testing-library/react'
import Tag from './index'

describe('Tag Component', () => {
  const mockOnClick = jest.fn()

  beforeEach(() => {
    mockOnClick.mockClear()
  })

  // SNAPSHOT TESTS FOR RENDERING
  describe('Rendering Snapshots', () => {
    it('renders button tag variants correctly', () => {
      // Default button tag
      const { container: defaultButton } = render(
        <Tag as="button" label="Button Tag" onClick={mockOnClick} />
      )
      expect(defaultButton).toMatchSnapshot('default-button')

      // Selected button tag
      const { container: selectedButton } = render(
        <Tag
          as="button"
          label="Selected Button"
          onClick={mockOnClick}
          variant="selected"
        />
      )
      expect(selectedButton).toMatchSnapshot('selected-button')

      // Disabled button tag
      const { container: disabledButton } = render(
        <Tag
          as="button"
          label="Disabled Button"
          onClick={mockOnClick}
          variant="disabled"
        />
      )
      expect(disabledButton).toMatchSnapshot('disabled-button')

      // Button with children instead of label
      const { container: childrenButton } = render(
        <Tag as="button" label="" onClick={mockOnClick}>
          <span>Custom Content</span>
        </Tag>
      )
      expect(childrenButton).toMatchSnapshot('button-with-children')
    })

    it('renders link tag variants correctly', () => {
      // Default link tag
      const { container: defaultLink } = render(
        <Tag as="link" label="Link Tag" href="/test" onClick={mockOnClick} />
      )
      expect(defaultLink).toMatchSnapshot('default-link')

      // Selected link tag
      const { container: selectedLink } = render(
        <Tag
          as="link"
          label="Selected Link"
          href="/test"
          onClick={mockOnClick}
          variant="selected"
        />
      )
      expect(selectedLink).toMatchSnapshot('selected-link')

      // Disabled link tag
      const { container: disabledLink } = render(
        <Tag
          as="link"
          label="Disabled Link"
          href="/test"
          onClick={mockOnClick}
          variant="disabled"
        />
      )
      expect(disabledLink).toMatchSnapshot('disabled-link')

      // Link with target blank
      const { container: targetLink } = render(
        <Tag
          as="link"
          label="External Link"
          href="/test"
          target="_blank"
          onClick={mockOnClick}
        />
      )
      expect(targetLink).toMatchSnapshot('link-with-target')
    })
  })

  // INTERACTIVE BEHAVIOR TESTS
  describe('Interactive Behavior', () => {
    it('calls onClick when button tag is clicked', () => {
      render(<Tag as="button" label="Button Tag" onClick={mockOnClick} />)

      fireEvent.click(screen.getByText('Button Tag'))
      expect(mockOnClick).toHaveBeenCalledTimes(1)
    })

    it('does not call onClick when disabled button tag is clicked', () => {
      render(
        <Tag
          as="button"
          label="Disabled Button"
          onClick={mockOnClick}
          variant="disabled"
        />
      )

      fireEvent.click(screen.getByText('Disabled Button'))
      expect(mockOnClick).not.toHaveBeenCalled()

      // Verify it's actually disabled
      expect(screen.getByRole('button')).toBeDisabled()
    })

    it('calls onClick when link tag is clicked', () => {
      render(
        <Tag as="link" label="Link Tag" href="/test" onClick={mockOnClick} />
      )

      fireEvent.click(screen.getByText('Link Tag'))
      expect(mockOnClick).toHaveBeenCalledTimes(1)
    })
  })
})
