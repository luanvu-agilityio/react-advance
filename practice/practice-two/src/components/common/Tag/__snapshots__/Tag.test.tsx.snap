// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Component Rendering Snapshots renders button tag variants correctly: button-with-children 1`] = `
<div>
  <button
    class="sc-fYsHOw kuAQwy"
    type="button"
  >
    <span>
      Custom Content
    </span>
  </button>
</div>
`;

exports[`Tag Component Rendering Snapshots renders button tag variants correctly: default-button 1`] = `
<div>
  <button
    class="sc-fYsHOw kuAQwy"
    type="button"
  >
    Button Tag
  </button>
</div>
`;

exports[`Tag Component Rendering Snapshots renders button tag variants correctly: disabled-button 1`] = `
<div>
  <button
    class="sc-fYsHOw dhNhzg"
    disabled=""
    type="button"
  >
    Disabled Button
  </button>
</div>
`;

exports[`Tag Component Rendering Snapshots renders button tag variants correctly: selected-button 1`] = `
<div>
  <button
    class="sc-fYsHOw cMHZiV"
    type="button"
  >
    Selected Button
  </button>
</div>
`;

exports[`Tag Component Rendering Snapshots renders link tag variants correctly: default-link 1`] = `
<div>
  <span
    class="sc-dKREkF bRrHLa"
  >
    <a
      class="sc-Qotzb jUyYcI"
      href="/test"
      target="_self"
    >
      Link Tag
    </a>
  </span>
</div>
`;

exports[`Tag Component Rendering Snapshots renders link tag variants correctly: disabled-link 1`] = `
<div>
  <span
    class="sc-dKREkF iSTpNP"
  >
    <a
      class="sc-Qotzb kzgBsp"
      disabled=""
      href="/test"
      target="_self"
    >
      Disabled Link
    </a>
  </span>
</div>
`;

exports[`Tag Component Rendering Snapshots renders link tag variants correctly: link-with-target 1`] = `
<div>
  <span
    class="sc-dKREkF bRrHLa"
  >
    <a
      class="sc-Qotzb jUyYcI"
      href="/test"
      rel="noopener noreferrer"
      target="_blank"
    >
      External Link
    </a>
  </span>
</div>
`;

exports[`Tag Component Rendering Snapshots renders link tag variants correctly: selected-link 1`] = `
<div>
  <span
    class="sc-dKREkF fdeVZx"
  >
    <a
      class="sc-Qotzb jUyYcI"
      href="/test"
      target="_self"
    >
      Selected Link
    </a>
  </span>
</div>
`;
